libraries/AP_Math/matrixN.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ *  Library for operations on matrices of arbitrary dimensions.
+ *
+ *  Based on the MatrixMath library created by Charlie Matlack on 12/18/10.
+ *  Modified from code by RobH45345 on Arduino Forums, taken from unknown source.
+ */
+
+#pragma GCC optimize("O3")
+ 
+#include "matrixN.h"
+
+void matrix_copy(float* A, int32_t n, int32_t m, float* B){
+    int32_t i, j;
+    
+    for (i = 0; i < m; i++){
+        for (j = 0; j < n; j++){
+            B[n * i + j] = A[n * i + j];
+        }
+    }
+}
+
+    
+// Matrix Multiplication Routine
+// C = A * B
+void matrix_mult(float* A, float* B, int32_t m, int32_t p, int32_t n, float* C){
```

Comment:
these matrix functions would be much nicer as a template class, and would probably be faster as you could avoid the index multiplies (ie. declare as fixed size 2D array)